### DISCUSSION TITLE
How to use ccache with vcpkg

### DISCUSSION THREAD

**Discussion by xiaozhuai** [microsoft/vcpkg#30191] [Q&A]
I upgrade my ports everyday, which waste a lot of time.
It's there any tricks that I can use ccache on building vcpkg ports?  (I use macos and windows msvc)
This will help me a lot. Thanks!

**Comment by klalumiere** [comment]
I've just tried and it works (I tried with [sccache](https://github.com/mozilla/sccache) but it shouln't change anything). :tada:  You need to set the environment variable [CMAKE_\<LANG\>_COMPILER_LAUNCHER](https://cmake.org/cmake/help/v3.20/envvar/CMAKE_LANG_COMPILER_LAUNCHER.html#envvar:CMAKE_%3CLANG%3E_COMPILER_LAUNCHER) to `ccache`, and if you are on Windows, you'll probably need to set [VCPKG_KEEP_ENV_VARS](https://learn.microsoft.com/en-us/vcpkg/users/config-environment) too.

On Linux, my whole command line looked like

```
CMAKE_CXX_COMPILER_LAUNCHER=sccache CMAKE_C_COMPILER_LAUNCHER=sccache  VCPKG_BINARY_SOURCES=clear cmake -B build2 -S .
```

(as you can see, I set my environment variable inline). You can use `ccache --show-stats` to ensure it's working.

**Reply by dg0yt** [reply]
These variables are specifc to CMake. I doubt they reach port which use other build systems. 
vcpkg does forward `CMAKE_<lang>_COMPILER` for many build systems. So it might be more promising to make that variable point to a wrapper which launches the compiler via ccache.

**Comment by Tagman** [comment]
I had the experience that you might need to rebuild your toolchain for the changes to take effect.

**Reply by xiaozhuai** [reply]
Could you explain it in details. Thanks!

**Reply by Tagman** [reply]
Sure. When using the manifest mode (having a vcpkg.json file in your project and running cmake with -DCMAKE_TOOLCHAIN_FILE=<path>), you need to:

- remove the vcpkg_installed folder from your  build directory.
- remove packages folder from your vcpkg directory
- remove buildtrees folder from vcpkg directory (just to be sure)
- remove archives directory from %LOCALAPPDATA%/vcpkg

This might be a bit overkill, but will lead to all your dependencies being rebuilt. If you are using the classic mode (e.g. calling vcpkg.exe install ...) you need to remove the vcpkg_installed directory from your vcpkg directory.

I hope this helps. Feel free to ask if it doesn't. :)

**Reply by dg0yt** [reply]
I don't understand why you would want to force a rebuild of installed binary artifacts to enable the use of `ccache`. Maybe that's the answer to a different type of problem?

**Reply by xiaozhuai** [reply]
> I don't understand why you would want to force a rebuild of installed binary artifacts to enable the use of `ccache`. 

Because `ccache` cause vcpkg binary caching expired, I guess.

**Reply by dg0yt** [reply]
> > I don't understand why you would want to force a rebuild of installed binary artifacts to enable the use of `ccache`.
> 
> Because `ccache` cause vcpkg binary caching expired, I guess.

This doesn't really anwer *why* one should remove *everything*, *manually*. 
Without knowing the actual problem, it doesn't make sense to suggest such radical measures.

**Reply by Tagman** [reply]
> I don't understand why you would want to force a rebuild of installed binary artifacts to enable the use of `ccache`. Maybe that's the answer to a different type of problem?

When I was trying to get ccache to work, I had no hits whenever I was just rebuilding the project. I had to remove my complete build folder. This was a 1-2hr try like 4 months ago, not sure if all that was really necessary.

I agree with you, it's radical, but its not the end of the world to do a rebuild of your toolchain and project. But it also might not lead to success. I was trying to use ccache with vcpkg and in my actual project. For now I skipped solving the problem as time is short. But it will pop up again in some weeks, so hopefully I can give you an explanation and solution.

**Comment by congzhangzh** [comment]
@xiaozhuai the follow way may work for some vcpkg port,  some other failed!

Use the wrapper and set CC/CXX, don’t forget to make CC/CXX keep

https://github.com/congzhangzh/daily-scripts-4-us/tree/main#2-three-way-to-make-ccache-work-on-windows

https://learn.microsoft.com/en-us/vcpkg/users/config-environment#vcpkg_keep_env_vars

**Comment by tsondergaard** [comment]
I've created https://github.com/microsoft/vcpkg/issues/37689.
